### PR TITLE
Suggest surrounding the macro with `{}` to interpret as a statement

### LINKED ISSUE #109237: Useless suggestion for macro expansion starting with `;` used in expression position
### Code

```Rust
/*
[package]
edition = "2021"
*/

macro_rules! statement {
    () => {;};
}

fn main() {
    let _ = statement!();
}
```

### Current output

```Shell
error: expected expression, found `;`
  --> src/main.rs:7:12
   |
7  |     () => {;};
   |            ^ expected expression
...
11 |     let _ = statement!();
   |             ------------ in this macro invocation
   |
   = note: the macro call doesn't expand to an expression, but it can expand to a statement
   = note: this error originates in the macro `statement` (in Nightly builds, run with -Z macro-backtrace for more info)
help: add `;` to interpret the expansion as a statement
   |
11 |     let _ = statement!();;
   |                         +
```

### Desired output

```Shell
error: expected expression, found `;`
  --> src/main.rs:7:12
   |
7  |     () => {;};
   |            ^ expected expression
...
11 |     let _ = statement!();
   |             ------------ in this macro invocation
   |
   = note: the macro call doesn't expand to an expression, but it can expand to a statement
   = note: this error originates in the macro `statement` (in Nightly builds, run with -Z macro-backtrace for more info)
```

or *perhaps*

```Shell
error: expected expression, found `;`
  --> src/main.rs:7:12
   |
7  |     () => {;};
   |            ^ expected expression
...
11 |     let _ = statement!();
   |             ------------ in this macro invocation
   |
   = note: the macro call doesn't expand to an expression, but it can expand to a statement
   = note: this error originates in the macro `statement` (in Nightly builds, run with -Z macro-backtrace for more info)
help: surround the macro with `{}` to interpret the expansion as a statement
   |
11 |     let _ = { statement!(); };
   |             ++            +++
```

### Rationale and extra context

The bad help isn't related to the choice of `m!(...)` or `m!{...}` (even though the former needs to be followed by `;` to be a statement and the latter doesn't, which could explain the source of the hint); `m!{...}` also generates the help to add `;`.

[[rustexplorer]](https://www.rustexplorer.com/b#%2F*%0A%5Bpackage%5D%0Aedition%20%3D%20%222021%22%0A*%2F%0A%0Amacro_rules!%20statement%20%7B%0A%20%20%20%20()%20%3D%3E%20%7B%3B%7D%3B%0A%7D%0A%0Afn%20main()%20%7B%0A%20%20%20%20let%20_%20%3D%20statement!()%3B%0A%7D%0A)

### Other cases

_No response_

### Anything else?

```Shell
$ rustc -Vv
rustc 1.70.0-nightly (ab654863c 2023-03-15)
binary: rustc
commit-hash: ab654863c3d50482f260cf862647f1fe0ff5e010
commit-date: 2023-03-15
host: x86_64-pc-windows-msvc
release: 1.70.0-nightly
LLVM version: 15.0.7
```

@rustbot modify labels +D-incorrect

### PR DESCRIPTION
**Pull request by mu001999**
Fixes #109237

### CODE REVIEW

File: compiler/rustc_expand/src/mbe/diagnostics.rs [hunk 1]
```diff
@@ -245,12 +245,24 @@ pub(super) fn emit_frag_parse_err(
                 e.note(
                     "the macro call doesn't expand to an expression, but it can expand to a statement",
```
Comment (Nilstrieb) [inline compiler/rustc_expand/src/mbe/diagnostics.rs:246]: The note here is correct of the old case, but not for your new code path. It's not that we want to turn an expression into a statement, but a statement into an expression.

Comment (MU001999) [reply compiler/rustc_expand/src/mbe/diagnostics.rs:246]: We are still turning a macro call being an expression (`statement!()` in the `let _ = statement!();`) into a macro call being a statement (`statement!();` in the whole `{...}`), although the whole `{..}` is an expression.

Comment (Nilstrieb) [reply compiler/rustc_expand/src/mbe/diagnostics.rs:246]: oh, right

**Comment by Nilstrieb** [comment]
r? Nilstrieb
@bors r+ rollup

**Pull request merged**